src/iosrv/drpc_listener.c

```diff
@@ -87,13 +86,13 @@ drpc_listener_run(void *arg)
 	while (is_listener_running()) {
 		int rc;
 
-		/* instant timeout - don't hog the xstream */
+		/* instant timeout */
 		rc = drpc_progress(ctx, 0);
 		if (rc != DER_SUCCESS && rc != -DER_TIMEDOUT) {
 			D_ERROR("dRPC listener progress error: %d\n", rc);
 		}
 
-		dss_sleep(SLEEP_TIME_MS); /* yield for a bit */
+		ABT_thread_yield();
```

Comment:
The zero timeout and the yield call defeat the purpose of a dedicated thread, don't they? We would like the listener to block indefinitely (or a couple seconds, if needed) in the poll call, while binding it to the same CPU as the system xstream.

Reply:
There was some bad interaction with the indefinite blocking that was hanging the iosrv during CI tests. But I haven't tried it since binding to the same CPU as the system xstream. I will retry and see how it goes.

Reply:
I did a little experiment with shortening the timeout to 1s (still had CaRT timeouts), then re-adding the yield (everything passes). Still not sure why yielding would have any effect. Every XS has a daos_srv_handler ULT, which progresses CaRT, but since we don't have CaRT comms on this XS, it shouldn't be doing anything. It's puzzling, I wasn't able to reproduce the issue on a 5-node config myself, but it happens in CI every time.

Reply:
I don't think it's a coincidence at this point--it always passes CI with the yield, and never passes without it--though it could be that this PR simply exposes a bug somewhere else.

Reply:
> The zero timeout and the yield call defeat the purpose of a dedicated thread, don't they? We would like the listener to block indefinitely (or a couple seconds, if needed) in the poll call, while binding it to the same CPU as the system xstream.

In this patch it changed to "rc = drpc_progress(ctx, 1000)" so it will block the drpc XS for 1 second if no drpc activity, and don't affect system XS although they are on same core. This seems fine?

I don't quite understand your comment though, you mean the "ABT_thread_yield()" call defeat the purpose of a dedicated thread?


as for why "dss_sleep(SLEEP_TIME_MS); /* yield for a bit */" cause the CI timeout, I am not sure, that seems will cause more conflict between drpc XS and system XS? (since in SLEEP_TIME_MS the ABT will need to do tons of yield + schedule back, and the OS need to frequent schedule between drpc XS/thread and system XS/thread on same core). But seems should not cause timeout commonly.

Reply:
> as for why "dss_sleep(SLEEP_TIME_MS); /* yield for a bit */" cause the CI timeout, I am not sure, that seems will cause more conflict between drpc XS and system XS? (since in SLEEP_TIME_MS the ABT will need to do tons of yield + schedule back, and the OS need to frequent schedule between drpc XS/thread and system XS/thread on same core). But seems should not cause timeout commonly.

Sorry for the confusion, it wasn't the dss_sleep causing the timeout. It was when we had no yields at all. I.e. The same code as the current PR version, but with the ABT_yield call removed. This version, for example, saw the timeouts: https://github.com/daos-stack/daos/pull/916/files/6901bba0934bbf24fd2be5d5095f9b5255f14290